### PR TITLE
[KB-204] Stacked Generalization

### PR DESCRIPTION
**Pull request by Ivikhostrup**
Reading guide:
* Is it okay to mix terms such as base model vs level-0 generalizers (this is the term used in the paper)
* I am unsure about the very last paragraph. I think its a nice summary, but I can also see why it might be redundant.

### CODE REVIEW

File: report_thesis/src/sections/background.tex [hunk 1]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
 
 \subsubsection{XGBoost}
 
-\subsection{Stacking Ensemble}
-% \citet{pavlyshenko2018stacking}
+\subsubsection{Stacked Generalization}
+Stacked generalization, introduced by \citet{wolpertstacked_1992}, is a technique designed to improve the predictive performance of machine learning models by leveraging the strengths of multiple models.
+
+In stacked generalization, multiple base models (level-0 generalizers) are trained on the original dataset.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:366]:
```suggestion
In this technique, multiple predictors are trained on the original dataset. 
```

File: report_thesis/src/sections/background.tex [hunk 2]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
 
 \subsubsection{XGBoost}
 
-\subsection{Stacking Ensemble}
-% \citet{pavlyshenko2018stacking}
+\subsubsection{Stacked Generalization}
+Stacked generalization, introduced by \citet{wolpertstacked_1992}, is a technique designed to improve the predictive performance of machine learning models by leveraging the strengths of multiple models.
+
+In stacked generalization, multiple base models (level-0 generalizers) are trained on the original dataset. 
+The predictions of these base models are then used as inputs to a meta-model (level-1 generalizer), which is trained to make the final prediction.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:367]:
```suggestion
The outputs of these base models serve as inputs to a meta-model, which is trained to make the final prediction.
```

File: report_thesis/src/sections/background.tex [hunk 3]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
 
 \subsubsection{XGBoost}
 
-\subsection{Stacking Ensemble}
-% \citet{pavlyshenko2018stacking}
+\subsubsection{Stacked Generalization}
+Stacked generalization, introduced by \citet{wolpertstacked_1992}, is a technique designed to improve the predictive performance of machine learning models by leveraging the strengths of multiple models.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:364]:
```suggestion
Stacked generalization, introduced by \citet{wolpertstacked_1992}, is a method designed to improve the predictive performance of machine learning models by leveraging the strengths of multiple models.
```

File: report_thesis/src/sections/background.tex [hunk 4]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
...
+
+In stacked generalization, multiple base models (level-0 generalizers) are trained on the original dataset. 
+The predictions of these base models are then used as inputs to a meta-model (level-1 generalizer), which is trained to make the final prediction. 
+This approach allows the meta-model to learn how to best combine the outputs of the base models to minimize generalization error.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:368]:
```suggestion
This strategy enables the meta-model to learn the optimal way to combine the outputs of the base models to minimize the generalization error.
```

File: report_thesis/src/sections/background.tex [hunk 5]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
...
+
+To formalize, let $\mathbf{X}$ be the input data and $\mathbf{y}$ be the target variable. 
+First, $N$ base models $G_1, G_2, \ldots, G_N$ are trained on the dataset $\mathbf{X}$. 
+Each base model produces a set of predictions $\hat{\mathbf{y}}_i = G_i(\mathbf{X})$.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:372]:
```suggestion
Formally, let $\mathbf{X}$ denote the input data and $\mathbf{y}$ the target variable.
Initially, $N$ base models $G_1, G_2, \ldots, G_N$ are trained on the dataset $\mathbf{X}$.
Each base model generates a set of predictions $\hat{\mathbf{y}}_i = G_i(\mathbf{X})$.
```

File: report_thesis/src/sections/background.tex [hunk 6]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
...
+
+The predictions from the base models are then combined into a new dataset $\mathbf{Z}$, where each column $\mathbf{z}_i \in \mathbf{Z}$ corresponds to the predictions from the $i$-th base model:
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:374]:
```suggestion
The predictions from the base models are then compiled into a new dataset $\mathbf{Z}$, where each column $\mathbf{z}_i \in \mathbf{Z}$ represents the predictions from the $i$-th base model:
```

File: report_thesis/src/sections/background.tex [hunk 7]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
...
+
+$$
+\mathbf{Z} = [\hat{\mathbf{y}}_1, \hat{\mathbf{y}}_2, \ldots, \hat{\mathbf{y}}_N]
+$$
+
+A meta-model $F$ is then trained on this new dataset $\mathbf{Z}$ to predict the target variable $\mathbf{y}$:
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:380]:
```suggestion
A meta-model $F$ is subsequently trained on this new dataset $\mathbf{Z}$ to predict the target variable $\mathbf{y}$:
```

File: report_thesis/src/sections/background.tex [hunk 8]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
...
+
+$$
+\mathbf{\hat{y}} = F(\mathbf{Z})
+$$
+
+The final prediction is given by the meta-model $F$, which effectively combines the outputs of the base models in an optimal way to improve overall performance.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:386]:
```suggestion
The final prediction is provided by the meta-model $F$, which effectively integrates the outputs of the base models to enhance overall performance.
```

File: report_thesis/src/sections/background.tex [hunk 9]
```diff
@@ -360,5 +360,29 @@ \subsubsection{Gradient Boosting Regression (GBR)}\label{sec:gradientboost}
...
+
+A meta-model $F$ is then trained on this new dataset $\mathbf{Z}$ to predict the target variable $\mathbf{y}$:
+
+$$
+\mathbf{\hat{y}} = F(\mathbf{Z})
+$$
+
+The final prediction is given by the meta-model $F$, which effectively combines the outputs of the base models in an optimal way to improve overall performance. 
+
+The power of stacked generalization lies in its ability to exploit the unique strengths of different base models and mitigate their individual weaknesses, leading to a more accurate and generalizable ensemble model.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:388]:
```suggestion
The efficacy of stacked generalization lies in its capability to leverage the distinct strengths of various base models while compensating for their individual weaknesses, resulting in a more accurate and generalizable ensemble model.
```

**Review by chhoumann — COMMENTED** [review]
Huge kudos for reading & using the original paper!
This section flows well, and I think your method of communication was great.
My comments are only about word choice - it's a little too informal for a thesis. So despite there being 'so many comments,' I thought you did a great job explaining stacking!

> Is it okay to mix terms such as base model vs level-0 generalizers (this is the term used in the paper)

I think it's confusing - doesn't add anything.

> I am unsure about the very last paragraph. I think its a nice summary, but I can also see why it might be redundant.

Nicely highlights benefits, so I think it's good

**Review by chhoumann — APPROVED** [review]

File: report_thesis/src/sections/background.tex [hunk 10]
```diff
@@ -393,5 +393,29 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+
+A meta-model $F$ is subsequently trained on this new dataset $\mathbf{Z}$ to predict the target variable $\mathbf{y}$:
+
+$$
+\mathbf{\hat{y}} = F(\mathbf{Z})
+$$
+
+The final prediction is provided by the meta-model $F$, which effectively integrates the outputs of the base models to enhance overall performance.
+
+The efficacy of stacked generalization lies in its capability to leverage the distinct strengths of various base models while compensating for their individual weaknesses, resulting in a more accurate and generalizable ensemble model.
```
Comment (Pattrigue) [inline report_thesis/src/sections/background.tex:421]:
```suggestion
The effectiveness of stacked generalization stems from its ability to leverage the unique strengths of different base models while mitigating their individual weaknesses, thereby producing a more accurate and generalizable ensemble model.
```

**Pull request merged**